media_source: fix remote source rendering

philn opened the pull request:
The previous version of the source bin had the capsfilter after the
glcolorconvert element, triggering a caps negotiation issue. And without
videoscale element the capsfilter is not very useful anyway.

So now the video resize is performed before uploading to the GPU.

@adam-be can you test this on your side please?

superdump commented:
I don't think this will be acceptable. The point is that the scaling should be done on the GPU and not in software. It's very cheap to scale in GL compared to on a CPU and this is especially relevant on embedded devices.

philn commented:
I get your point. So something else is broken then, I'll investigate further!

philn committed:
media_source: fix remote source rendering

The previous version of the source bin had the capsfilter after the
glcolorconvert element, triggering a caps negotiation issue. The video
scaling is now properly done in the GPU using the gleffects_identity
element.

philn commented:
The patch is updated. Please review :)

superdump commented:
I guess this gleffects_identity element is a no-op...? If so, what is the point of having it at all - does it basically support scaling in GL? Would that not be done by the glupload element?

Also, I don't think the capsfilter should be removed. The purpose of the capsfilter is to specify the desired output format, which is why it follows the conversion elements.

What is the cause of the caps negotiation issue?

philn commented:
gleffects_identity isn't a no-op, it uses a shader to scale video textures depending on the caps. It's the only option I found to do scaling in the GPU. I tried glvideoscale and that failed. AFAIK glupload doesn't scale the textures it uploads to the GPU.

I removed the capsfilter simply because the output format is negotiated when connecting the source to the renderer where the caps are configured.

The initial caps negotiation issue happens when the remote owr media source connects to the renderer. The source has caps set on the video decoder to, say, 800x600 I420, but the renderer is configured with different caps, I'll post debug logs in a follow-up comment.

philn commented:
https://gist.github.com/philn/8dc56dace3cf29a09c1e259dc841b1a0

In the raw file look for "could not transform"

stefhak commented:
@superdump any further comments after this update?
